File: src/core/DepthFirstGraph.java
package core;

import java.util.HashSet;
import java.util.Set;
import java.util.Stack;
import java.util.stream.Collectors;

import path.State;

public class DepthFirstGraph<S extends State<S>> {
	private Stack<S> frontier;
	private Set<S> explored;
	
	public DepthFirstGraph(S startState) {
		frontier = new Stack<>();
		frontier.push(startState);
		
		explored = new HashSet<>();
	}
	
	public S run() {
		S currentState;
		do {
			currentState = frontier.pop();
			if (!explored.add(currentState)) {
				// This state was already contained in the explored set; ignore it.
				continue;
			}
			// Determine the available actions, and add each of their resulting states to the frontier.
			frontier.addAll(currentState.determineAvailableActions().stream().map(action -> action.getResultingState()).collect(Collectors.toList()));
		} while (!currentState.isGoalState());
		
		return currentState;
	}
	
}

File: src/main/scala/nl/calmamity/search/path/DepthFirstGraph.scala
package nl.calmamity.search.path

import core.Action

import scala.collection.mutable

/** Implements a graph-based form of the depth-first search algorithm. */
class DepthFirstGraph[StateImplementation <: State[StateImplementation]](startState: StateImplementation) {
	/** A collection of all nodes that have not yet been expanded. */
	val frontier: mutable.ArrayStack[StateImplementation] = new mutable.ArrayStack[StateImplementation]()
	frontier.push(startState)
	
	/** The states that have already been explored; if encountered again, states in this set will not be expanded a second time. */
	val explored: mutable.Set[StateImplementation] = mutable.Set[StateImplementation]()
	
	/**
	  * Runs the algorithm to completion.
	  * @return the goal state found by the algorithm.
	  */
	def run(): StateImplementation = {
		var currentState: StateImplementation = frontier.pop()
		
		while (!currentState.isGoalState) {
			if (explored.add(currentState)) {
				// Determine the available actions, and add each of their resulting states to the frontier.
				val availableActionsIterator = currentState.createAvailableActionsIterator()
				while (availableActionsIterator.hasNext) {
					frontier.push(availableActionsIterator.next().getResultingState)
				}
			}
			currentState = frontier.pop()
		}
		
		currentState
	}
}
